src/Token/Verifier.php

```diff
@@ -0,0 +1,234 @@
+            }
+
+            $key = $this->getKey($kid);
+            $valid = openssl_verify($this->payload, $this->signature, $key, OPENSSL_ALGO_SHA256);
+            $this->freeKey($key);
+
+            if ($valid !== 1) {
+                throw \Auth0\SDK\Exception\InvalidTokenException::badSignature();
+            }
+
+            return $this;
+        }
+
+        if ($alg === Token::ALGO_HS256) {
+            if (! $this->clientSecret) {
+                throw \Auth0\SDK\Exception\InvalidTokenException::requiresClientSecret();
+            }
+
+            $hash = hash_hmac('sha256', $this->payload, $this->clientSecret, true);
+            $valid = hash_equals($this->signature, $hash);
+
+            if (! $valid) {
+                throw \Auth0\SDK\Exception\InvalidTokenException::badSignature();
+            }
+
+            return $this;
+        }
+
+        throw \Auth0\SDK\Exception\InvalidTokenException::unsupportedSigningAlgorithm($alg);
+    }
+
+    /**
+     * Query a JWKS endpoint and return an array representing the key set.
+     *
+     * @throws InvalidTokenException When the JWKS uri is not properly configured, or is unreachable.
+     */
+    protected function getKeySet(): array
+    {
+        if ($this->jwksUri === null) {
+            throw \Auth0\SDK\Exception\InvalidTokenException::requiresJwksUri();
+        }
+
+        $jwksCacheKey = md5($this->jwksUri);
+        $jwksUri = parse_url($this->jwksUri);
+
+        $scheme = $jwksUri['scheme'] ?? 'https';
+        $path = $jwksUri['path'] ?? '/.well-known/jwks.json';
+        $response = [];
+
+        if ($this->cache !== null) {
+            $cached = $this->cache->get($jwksCacheKey);
+
+            if ($cached !== null) {
+                return $cached;
+            }
+        }
+
+        $keys = (new RequestBuilder([
+            'method' => 'get',
+            'domain' => $scheme . '://' . $jwksUri['host'],
+            'basePath' => $path,
+        ]))->call();
+
+        if (is_array($keys) && isset($keys['keys']) && count($keys['keys'])) {
+            foreach ($keys['keys'] as $key) {
+                if (isset($key['kid']) && isset($key['x5c']) && is_array($key['x5c']) && count($key['x5c'])) {
+                    $response[$key['kid']] = $key;
+                }
+            }
+        }
+
+        if (count($response) && $this->cache !== null) {
+            $this->cache->set($jwksCacheKey, $response, $this->cacheExpires ?? 60);
+        }
+
+        return $response;
+    }
+
+    /**
+     * Query a JWKS endpoint for a matching key. Parse and return a OpenSSLAsymmetricKey suitable for verification.
+     *
+     * @param string $kid The 'kid' header value to use for key lookup.
+     *
+     * @throws InvalidTokenException When unable to retrieve key. See error message for details.
+     */
+    protected function getKey(
+        string $kid
+    ): OpenSSLAsymmetricKey {
+        $keys = $this->getKeySet();
+
+        if (! isset($keys[$kid])) {
```

Comment:
Looks like the [old implementation](https://github.com/auth0/auth0-PHP/blob/78bd6673eef0bbc9c63c709bc374767ef89568aa/src/Helpers/JWKFetcher.php#L85-L87) fetched new keys if `kid` wasn't found in the cache and this one doesn't. But this one has a shorter TTL by default, I wonder if this is good enough to support Signing Key Rotation?

Ideally, if a key is not found, the cache should be invalidated and you should fetch a new JWKS document (with a cooldown period to prevent DOS)
https://auth0team.atlassian.net/wiki/spaces/DXSDK/pages/392563492/Signing+Key+Rotation+Area+of+Effect+Discovery#SigningKeyRotation:AreaofEffectDiscovery-CacheInvalidation

Reply:
Good call, that was definitely just an oversight on my part. I'll re-add that behavior.

Reply:
@adamjmcgrath I've updated this logic: Verifier::getKeySet() now accepts a parameter ($expectsKid) indicating the kid we're expecting to find. If there is a cached key set, but that cache doesn't contain that key, it will now auto-refresh the JWKS before returning the response, as if there had been no cached result at all.

Reply:
@evansims new behaviour lgtm 👍  just needs a regression test.

Also, I know it wasn't in the previous version and not required for this PR, but I would consider adding the cooldown behaviour in this major release or a minor shortly afterwards. (see [panva/jose/jwks](https://github.com/panva/jose/blob/main/src/jwks/remote.ts#L159) for an example implementation)

Reply:
Also, not for this PR, but another nice to have is to reuse the any pending request if the `kid` is not found.

If your app is under high qps, and it takes a a moment to fetch the new JWKS, you might spawn many new requests to the jwks endpoint, when you only really want one.

Not sure how this would work in PHP, but see [pendingfetch in panva/jose.jwks](https://github.com/panva/jose/blob/main/src/jwks/remote.ts#L189) for the node implementation

Reply:
@adamjmcgrath I've updated the test suite to cover this now; decided to go ahead and rewrite the Verifier unit tests to the new Pest format. (I'll be doing this with additional classes in later PRs.)

Thanks for the additional info, as well! Extremely helpful, and I agree we need to get those in there. Let me think about how we could approach that in PHP, there are a few challenges I need to think through there. I'd like to get it in with this next release for sure.